### DISCUSSION TITLE
Few questions about setting up the development environment

### DISCUSSION THREAD

**Discussion by Vafa-Andalibi** [tonaljs/tonal#328] [Q&A]
Hi, 

I'm new to typescript and am trying to add some functionality to some packages in this repo. I tried to to create a node/typescript package and add `tonaljs` as a dependency and develop my package while modifying the `tonaljs`. To achieve this, I tried to rely on `nodemon`  so that whenever I make any change on the dependency (@tonaljs), they are automatically propagated to my package. 

What I did was: 
- Create a new package (my main project)
- Build `tonaljs` using `yarn install` followed by `yarn build:all`
- Use `npm link` on each folder in `tonal/packages/` (I also tried this step on all built files in `tonal/node_modules/@tonaljs`
- Use `npm link <name>` to add each package from previous step to my own package
- [at this point the dependency works fine and I can use the library]
- Use `nodemon` on the main `tonaljs` folder (NOT the one in my package's `node_module` folder) to monitor changes (tried this on the `tonal/packages` folder as well as a specific folder like `tonal/packages/intervals`).

The problem is that `nodemon` detects changes, but the output of API doesn't change in my package.

Few questions:
- What am I doing wrong?
- Is there any easier way to do this?
- Why doesn't the `main` file from `package.json` exist before the build?
- If I want to create the main file on top of this library and modify the library while developing the main, can I do that right in the `tonal` folder, or do I have to create another project and add this as a dependency?

Thank you,

**Comment by danigb** [comment]
Hi! Not very tonal specific, but some hints here:

> What am I doing wrong?
Well... tonal (and npm modules in general) are no expected to be modified, but extended. In this way, you should be able to add functionality without changing the tonal modules themselves. That will simplify everything a lot 

So I don't have an answer. Maybe if you explain specifically what are you trying to achieve that requires tonal modification, I could help a little bit further

**Reply by Vafa-Andalibi** [reply]
For example, I'm trying to add quarter-tone additions ([example](https://en.wikipedia.org/wiki/Koron_(music))), and add oriental scales. That's my main goal.

**Comment by danigb** [comment]
Ah, that's great. Unfortunately tonal is not well suited for non-well-tempered tuning or scales, basically because it uses note names (or intervals, all in wester conception of the term) as source of truth. As a lover of music from all around the world, I've been thinking for a while to create something that supports other tunings or scales, but can't find time or motivation enough 🤷 

Anyway, you can add new scales using https://github.com/tonaljs/tonal/tree/main/packages/scale-type#addintervals-string-name-string-aliases-string--scaletype

**Reply by Vafa-Andalibi** [reply]
I see thanks for the answer! Again this is kind of an irrelevant question: when you are developing tonal, and you add something, how do you test it without rebuilding the library manually every time? Or do you do test-driven development?
